packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -3255,32 +3257,28 @@ export default (superClass: Class<Parser>): Class<Parser> =>
         }
         expr.typeParameters = typeParameters;
 
-        // report error if single type parameter used without trailing comma.
-        if (
-          this.hasPlugin("jsx") &&
-          expr.typeParameters.params.length === 1 &&
-          !expr.typeParameters.extra?.trailingComma
-        ) {
-          const parameter = expr.typeParameters.params[0];
-          if (!parameter.constraint) {
-            // A single type parameter must either have constraints
-            // or a trailing comma, otherwise it's ambiguous with JSX.
-            invalidSingleType = parameter;
+        if (process.env.BABEL_8_BREAKING) {
+          if (
+            this.hasPlugin("jsx") &&
+            expr.typeParameters.params.length === 1 &&
+            !expr.typeParameters.extra?.trailingComma
+          ) {
+            // report error if single type parameter used without trailing comma.
+            const parameter = expr.typeParameters.params[0];
+            if (!parameter.constraint) {
+              // A single type parameter must either have constraints
+              // or a trailing comma, otherwise it's ambiguous with JSX.
+              this.raise(TSErrors.SingleTypeParameterWithoutTrailingComma, {
```

Comment:
I moved this error within the `this.tryParse` call, so that it know that there is a problem with the function. Otherwise, when parsing `<T>() => {}</T>` it would give the precedence to the arrow function because it thought that `<T>() => {}` could be pared without errors.